torch/csrc/autograd/init.cpp

```diff
@@ -955,6 +955,120 @@ static PyObject* is_fwd_grad_enabled(PyObject* _unused, PyObject* arg) {
   END_HANDLE_TH_ERRORS
 }
 
+template <bool skip_tensors_in_non_tensorlist = false>
+static void visit(
+    PyObject* o,
+    const std::function<bool(at::Tensor&)>& visit_tensor) {
+  if (THPVariable_Check(o)) {
+    auto t = THPVariable_Unpack(o);
+    if (visit_tensor(t)) {
+      return;
+    }
+  } else if (PyList_Check(o)) {
+    // Check that this List is TensorList
+    if constexpr (skip_tensors_in_non_tensorlist) {
+      for (const auto i : c10::irange(PyList_GET_SIZE(o))) {
+        if (!THPVariable_Check(PyList_GET_ITEM(o, i))) {
+          return;
+        }
+      }
+    }
+    for (const auto i : c10::irange(PyList_GET_SIZE(o))) {
+      visit(PyList_GET_ITEM(o, i), visit_tensor);
+    }
+  }
+}
+
+template <bool skip_tensors_in_non_tensorlist = false>
+static void visit_tensors(
+    PyObject* args,
+    PyObject* kwargs,
+    const std::function<bool(at::Tensor&)>& visit_tensor) {
+  if (args && PyTuple_Check(args)) {
+    for (const auto i : c10::irange(PyTuple_GET_SIZE(args))) {
+      visit<skip_tensors_in_non_tensorlist>(
+          PyTuple_GET_ITEM(args, i), visit_tensor);
+    }
+  }
+  if (kwargs && PyDict_Check(kwargs)) {
+    auto vals = PyDict_Values(kwargs);
+    for (const auto i : c10::irange(PyList_GET_SIZE(vals))) {
+      visit<skip_tensors_in_non_tensorlist>(
+          PyList_GET_ITEM(vals, i), visit_tensor);
+    }
+  }
+}
```

Comment:
@IvanKobzarev this function doesn't short circuit, right? `any_requires_grad` calls visit_tensors. This is the function that I would expect to short circuit.